aws-storage-s3/src/main/java/com/amplifyframework/storage/s3/operation/AWSS3StoragePathDownloadFileOperation.kt

```diff
@@ -0,0 +1,255 @@
+ * permissions and limitations under the License.
+ */
+package com.amplifyframework.storage.s3.operation
+
+import com.amplifyframework.auth.AuthCredentialsProvider
+import com.amplifyframework.core.Amplify
+import com.amplifyframework.core.Consumer
+import com.amplifyframework.hub.HubChannel
+import com.amplifyframework.hub.HubEvent
+import com.amplifyframework.storage.IdentityIdProvidedStoragePath
+import com.amplifyframework.storage.StorageChannelEventName
+import com.amplifyframework.storage.StorageException
+import com.amplifyframework.storage.StoragePathValidationException
+import com.amplifyframework.storage.StringStoragePath
+import com.amplifyframework.storage.TransferState
+import com.amplifyframework.storage.operation.StorageDownloadFileOperation
+import com.amplifyframework.storage.result.StorageDownloadFileResult
+import com.amplifyframework.storage.result.StorageTransferProgress
+import com.amplifyframework.storage.s3.extensions.invalidStoragePathException
+import com.amplifyframework.storage.s3.extensions.unsupportedStoragePathException
+import com.amplifyframework.storage.s3.request.AWSS3StoragePathDownloadFileRequest
+import com.amplifyframework.storage.s3.service.StorageService
+import com.amplifyframework.storage.s3.transfer.TransferListener
+import com.amplifyframework.storage.s3.transfer.TransferObserver
+import java.io.File
+import java.util.UUID
+import java.util.concurrent.ExecutorService
+import kotlinx.coroutines.runBlocking
+
+/**
+ * An operation to download a file from AWS S3.
+ */
+internal class AWSS3StoragePathDownloadFileOperation(
+    private val transferId: String = UUID.randomUUID().toString(),
+    private val request: AWSS3StoragePathDownloadFileRequest?,
+    private var file: File,
+    private val storageService: StorageService,
+    private val executorService: ExecutorService,
+    private val authCredentialsProvider: AuthCredentialsProvider,
+    private var transferObserver: TransferObserver? = null,
+    onProgress: Consumer<StorageTransferProgress>? = null,
+    onSuccess: Consumer<StorageDownloadFileResult>? = null,
+    onError: Consumer<StorageException>? = null
+) :
+    StorageDownloadFileOperation<AWSS3StoragePathDownloadFileRequest>(
+        request,
+        transferId,
+        onProgress,
+        onSuccess,
+        onError
+    ) {
+
+    constructor(
+        request: AWSS3StoragePathDownloadFileRequest,
+        storageService: StorageService,
+        executorService: ExecutorService,
+        authCredentialsProvider: AuthCredentialsProvider,
+        onProgress: Consumer<StorageTransferProgress>? = null,
+        onSuccess: Consumer<StorageDownloadFileResult>? = null,
+        onError: Consumer<StorageException>? = null
+    ) : this(
+        UUID.randomUUID().toString(),
+        request,
+        request.local,
+        storageService,
+        executorService,
+        authCredentialsProvider,
+        null,
```

Comment:
what's the use case for a null transferobserver? 🤔

Reply:
TransferObserver does not come back until the transfer is being started which will happen after some async work.